drivers/kscan/kscan_cst816s.c

```diff
@@ -0,0 +1,333 @@
+
+	if (!callback) {
+		LOG_ERR("Invalid callback (NULL)");
+		return -EINVAL;
+	}
+
+	data->callback = callback;
+
+	return 0;
+}
+
+static int cst816s_enable_callback(const struct device *dev)
+{
+	struct cst816s_data *data = dev->data;
+
+#ifdef CONFIG_KSCAN_CST816S_INTERRUPT
+	const struct cst816s_config *config = dev->config;
+	gpio_add_callback(config->int_gpio.port, &data->int_gpio_cb);
+#else
+	k_timer_start(&data->timer, K_MSEC(CONFIG_KSCAN_cst816s_PERIOD),
+		      K_MSEC(CONFIG_KSCAN_cst816s_PERIOD));
+#endif
+
+	return 0;
+}
+
+static int cst816s_disable_callback(const struct device *dev)
+{
+	struct cst816s_data *data = dev->data;
+
+#ifdef CONFIG_KSCAN_CST816S_INTERRUPT
+	const struct cst816s_config *config = dev->config;
+	gpio_remove_callback(config->int_gpio.port, &data->int_gpio_cb);
+#else
+	k_timer_stop(&data->timer);
+#endif
+
+	return 0;
+}
+
+#if DT_INST_NODE_HAS_PROP(0, rst_gpios)
+static void cst816s_chip_reset(const struct device *dev)
+{
+	const struct cst816s_config *cfg = dev->config;
+
+	gpio_pin_set_dt(&cfg->rst_gpio, 1);
+	k_msleep(5);
+	gpio_pin_set_dt(&cfg->rst_gpio, 0);
+	k_msleep(50);
+}
+#endif
+
+static int cst816s_chip_init(const struct device *dev)
+{
+	const struct cst816s_config *cfg = dev->config;
+	int ret = 0;
+	uint8_t chip_id = 0;
+
+#if DT_INST_NODE_HAS_PROP(0, rst_gpios)
+	cst816s_chip_reset(dev);
+#endif
+	if (!device_is_ready(cfg->i2c.bus)) {
+		LOG_ERR("I2C bus %s not ready", cfg->i2c.bus->name);
+		return -ENODEV;
+	}
+	ret = i2c_reg_read_byte_dt(&cfg->i2c,
+				CST816S_REG_CHIP_ID, &chip_id);
+	if (ret < 0) {
+		LOG_ERR("failed reading chip id");
+		return ret;
+	}
+
+	if (chip_id != CST816S_CHIP_ID) {
+		LOG_ERR("CST816S wrong chip id: returned 0x%x", chip_id);
+		return -ENODEV;
+	}
+
+	ret = i2c_reg_update_byte_dt(&cfg->i2c,
+				CST816S_REG_IRQ_CTL,
+				CST816S_IRQ_EN_TOUCH | CST816S_IRQ_EN_CHANGE,
+				CST816S_IRQ_EN_TOUCH | CST816S_IRQ_EN_CHANGE);
+	if (ret < 0) {
+		LOG_ERR("Could not enable irq");
+		return ret;
+	}
+	return ret;
+}
+
+static int cst816s_init(const struct device *dev)
+{
+	const struct cst816s_config *config = dev->config;
+	struct cst816s_data *data = dev->data;
+
+	data->dev = dev;
+	k_work_init(&data->work, cst816s_work_handler);
+
+#ifdef CONFIG_KSCAN_CST816S_INTERRUPT
+	int r;
+
+	if (!device_is_ready(config->int_gpio.port)) {
+		LOG_ERR("GPIO port %s not ready", config->int_gpio.port->name);
+		return -EINVAL;
+	}
+
+	r = gpio_pin_configure_dt(&config->int_gpio, GPIO_INPUT);
+	if (r < 0) {
+		LOG_ERR("Could not configure interrupt GPIO pin");
+		return r;
+	}
+
+	r = gpio_pin_interrupt_configure_dt(&config->int_gpio,
+			GPIO_INT_EDGE_TO_ACTIVE);
+	if (r < 0) {
+		LOG_ERR("Could not configure interrupt GPIO interrupt.");
+		return r;
+	}
+
+	gpio_init_callback(&data->int_gpio_cb, cst816s_isr_handler,
+			   BIT(config->int_gpio.pin));
+#else
+	k_timer_init(&data->timer, cst816s_timer_handler, NULL);
+#endif
+
+#if DT_INST_NODE_HAS_PROP(0, rst_gpios)
```

Comment:
this prevents driver from being multi-instance, what if instance 1 has rst gpios?